### DISCUSSION TITLE
why not always use libhdfs3, but use libhdfs3 for x86

### DISCUSSION THREAD

**Discussion by josehu-t** [apache/doris#45790] [A - General / Q&A] [answered]
The commit below introduces libhdfs, but not kown the purpose of it. Does it related with that libhdfs3 can not connect to HADOOP-COS which extends HDFS.  libhdfs.so operates on the principle of JNI, where C++ calls the Hadoop filesystem interface. Since COSN encapsulates the Hadoop filesystem interface, it connects to COSN. libhdfs3 uses the private HDFS Protocol Buffers (pb) protocol. Is there a better way to solve this. We don't want jni. Thanks advance for anyone help.

https://github.com/apache/doris/commit/7e61a8533151d80a903d2452119b77bca59b43bc

**Comment by morningman** [comment, marked as answer]
The reason we introduce hadoop libhdfs instead of libhdfs3 just because we can not address the case you mentioned, to access various hadoop-compatible file system like COSN.
libhdfs3 missed too many features that can not work well with hadoop eco system.
Currently, there is no other solution.
